# Distribuição de Poisson

# Uma central de monitoramento recebe eventos de falha em um sistema IoT de sensores agrícolas.

**O número médio de falhas observadas é 4 falhas por hora.**

a) Qual é a probabilidade de ocorrerem exatamente 6 falhas em uma hora?  
b) Qual é a probabilidade de ocorrerem no máximo 2 falhas em uma hora?  
c) Qual é a probabilidade de ocorrerem mais de 5 falhas em duas horas?  
d) Calcule o valor esperado e a variância da variável aleatória associada.  

<a href="https://colab.research.google.com/github/jgocana/Trabalhos-TP-547/blob/master/Trabalho%201/distribuição_de_poisson.ipynb" target="_parent">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [ ]:
!pip install numpy
!pip install math

Note: you may need to restart the kernel to use updated packages.


In [1]:
"""
Distribuição de Poisson

Este script resolve os incisos (a), (b), (c) e (d) utilizando:
1) Cálculo teórico da distribuição de Poisson
2) Simulação via método da inversa

Parâmetro:
λ = 4 falhas por hora
"""

import numpy as np
import math

# ==============================
# Parâmetros
# ==============================
lambda1 = 4   # média de falhas por hora
N = 100000    # número de amostras para simulação

# ==============================
# Função PMF de Poisson (teórica)
# ==============================
def poisson_pmf(k, lam):
    return (lam**k * np.exp(-lam)) / math.factorial(k)

# ==============================
# Simulação usando método da inversa
# ==============================
def gerar_poisson(lam, N):
    amostras = []
    u = np.random.uniform(0, 1, N)
    
    for ix in u:
        i = 0
        pr = np.exp(-lam)
        F = pr
        
        while ix >= F:
            pr = lam/(i+1) * pr
            F += pr
            i += 1
        
        amostras.append(i)
    
    return np.array(amostras)

# Gerar amostras
amostras = gerar_poisson(lambda1, N)

# ==============================
# a) P(X = 6)
# ==============================
p_teorico_a = poisson_pmf(6, lambda1)
p_simulado_a = np.mean(amostras == 6)

print("a) P(X = 6)")
print("Teórico:", p_teorico_a)
print("Simulado:", p_simulado_a)

# ==============================
# b) P(X ≤ 2)
# ==============================
p_teorico_b = sum(poisson_pmf(k, lambda1) for k in range(3))
p_simulado_b = np.mean(amostras <= 2)

print("\nb) P(X ≤ 2)")
print("Teórico:", p_teorico_b)
print("Simulado:", p_simulado_b)

# ==============================
# c) P(X > 5) em 2 horas
# ==============================
# Em 2 horas: λ = 8
lambda2 = 8

# Simulação para 2 horas
amostras_2h = gerar_poisson(lambda2, N)

p_teorico_c = 1 - sum(poisson_pmf(k, lambda2) for k in range(6))
p_simulado_c = np.mean(amostras_2h > 5)

print("\nc) P(X > 5) em 2 horas")
print("Teórico:", p_teorico_c)
print("Simulado:", p_simulado_c)

# ==============================
# d) Valor esperado e variância
# ==============================
# Teórico
esperado_teorico = lambda1
variancia_teorica = lambda1

# Simulado
esperado_simulado = np.mean(amostras)
variancia_simulada = np.var(amostras)

print("\nd) Valor esperado e variância")
print("Esperado teórico:", esperado_teorico)
print("Esperado simulado:", esperado_simulado)
print("Variância teórica:", variancia_teorica)
print("Variância simulada:", variancia_simulada)



a) P(X = 6)
Teórico: 0.1041956345670211
Simulado: 0.10394

b) P(X ≤ 2)
Teórico: 0.2381033055535443
Simulado: 0.23918

c) P(X > 5) em 2 horas
Teórico: 0.8087639379203747
Simulado: 0.80707

d) Valor esperado e variância
Esperado teórico: 4
Esperado simulado: 3.99849
Variância teórica: 4
Variância simulada: 3.9941877198999998


## Resultados  
---  

### a) Probabilidade de exatamente 6 falhas

\[
P(X = 6)
\]

- Teórico: 0.1041956345670211  
- Simulado: 0.10394

---

### b) Probabilidade de no máximo 2 falhas

\[
P(X ≤ 2)
\]

- Teórico: 0.2381033055535443  
- Simulado: 0.23969  

---

### c) Probabilidade de mais de 5 falhas em 2 horas

\[
P(X > 5), λ = 8
\]

- Teórico: 0.8087639379203747  
- Simulado: 0.80853  

---

### d) Valor esperado e variância

- Valor esperado teórico: 4  
- Valor esperado simulado: 3.99246  

- Variância teórica: 4  
- Variância simulada: 4.0074831484  

